### DISCUSSION TITLE
OpenTelemetry Tracing support for PowerDNS Products?

### DISCUSSION THREAD

**Discussion by dmachard** [PowerDNS/pdns#14992] [Ideas]
Hello,

I am wondering if there are any plans to integrate [OpenTelemetry](https://opentelemetry.io/  ) support into your PowerDNS products (such as dnsdist and Recursor). I believe that adding tracing functionality could be extremely useful for debugging complex DNS situations.

For context, here’s an overview of an existing implementation using [DNS-collector](https://github.com/dmachard/go-dnscollector), your products (dnsdist/recursor), and [Tempo](https://grafana.com/oss/tempo/) from Grafana. In my environment, DNSdist is deployed behind Recursor, and tracing allows me to visualize the path traversal through each components (dnsdist first, then recursor).

![otel dnscollector](https://github.com/user-attachments/assets/8fa22298-2c50-4204-8445-34059364dcff)

As shown, we can use Grafana to visualize relationships between DNS queries and replies, which is particularly helpful in troubleshooting DNS errors.

![otel dnscollector error](https://github.com/user-attachments/assets/8a026569-d45f-4835-8738-64c6a3146cb1)

I would like to point out that while DNS-collector currently handles tracing, it may not be the most efficient approach. It would be significantly more effective if tracing was implemented directly within each product (dnsdist and Recursor) and if they were able to share OpenTelemetry context between them.

I believe this feature could greatly enhance the debugging and monitoring capabilities for DNS infrastructures.

Denis

**Comment by neilcook** [comment]
Hi Denis, yes we are in fact looking at exactly this on the roadmap.
